### ISSUE TITLE
pywalfox-native: `pywalfox install` fails

### ISSUE THREAD

**Issue by auctumnus** [NixOS/nixpkgs#281377]
### Describe the bug
The package `pywalfox-native` requires an "install" step which is not currently in the package. Attempting to perform this imperatively fails:

```
autumn@birch ~> pywalfox install
Successfully removed manifest at: /home/autumn/.mozilla/native-messaging-hosts/pywalfox.json
Copied manifest to: /home/autumn/.mozilla/native-messaging-hosts/pywalfox.json
Set daemon executable path to: /nix/store/3al04ncj4haz5hxm7jj1b1mvzbz5dmk1-pywalfox-native-2.7.4/lib/python3.11/site-packages/pywalfox/bin/main.sh
Failed to set executable permissions on: /nix/store/3al04ncj4haz5hxm7jj1b1mvzbz5dmk1-pywalfox-native-2.7.4/lib/python3.11/site-packages/pywalfox/bin/main.sh
	[Errno 30] Read-only file system: '/nix/store/3al04ncj4haz5hxm7jj1b1mvzbz5dmk1-pywalfox-native-2.7.4/lib/python3.11/site-packages/pywalfox/bin/main.sh'

Try setting the permissions manually using: chmod +x
```

### Steps To Reproduce
1. Install `pywalfox-native`.
2. Run `pywalfox install`.

### Expected behavior
Theoretically, this should be run for the user, without the user needing to run it. I'm not sure what `pywalfox install` needs, exactly, but in any case `pywalfox install` needs to be able to be run at some point.

### Notify maintainers

@tsandrini 

### Metadata
```console
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.68, NixOS, 23.11 (Tapir), 23.11.20231220.d65bcea`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```
---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by tsandrini** [comment]
@auctumnus Hi! The pywalfox install isn't exactly nix/NixOS friendly + imperative in nature ... and since the only important part that it does is copying a manifest file I usually just do it myself. For example here is my personal home-manager module for inspiration https://github.com/tsandrini/tensorfiles/blob/main/parts/modules/home-manager/services/pywalfox-native.nix 

I'm not exactly sure how to cover every case since for some people the path for the manifest is `/usr/lib/mozilla/native-messaging-hosts` (root install) and `.mozilla/native-messaging-hosts` for others that either only use home-manager or simply want it installed in the userspace only, which is inaccessible to simple nix setups. I could try simply putting it in the former path and see if it works for the latter mentioned cases I guess?

**Comment by DaniD3v** [comment]
There's a branch on the pywalfox repo that works flawlessly on nixOS

read more: https://github.com/Frewacom/pywalfox/issues/68

**Comment by tsandrini** [comment]
Should be resolved, let me know if any issues persist.

**Comment by DaniD3v** [comment]
Could you backport this to stable?

**Comment by Fxzzi** [comment]
Hey there,

I'm having pywalfox also fail to work after moving from Arch Linux. I use the librewolf browser as I did on arch, and I was using 2.8.0rc1 like I did there as well.

It seems to install just fine as shown here:

```
Using executable path: /nix/store/djix9cnscraack2iawpindz9lfj27mdy-pywalfox-native-2.7.4/bin/pywalfox
Successfully removed manifest at: /home/faaris/.librewolf/native-messaging-hosts/pywalfox.json
Copied manifest to: /home/faaris/.librewolf/native-messaging-hosts/pywalfox.json
Set executable path in manifest to: /nix/store/djix9cnscraack2iawpindz9lfj27mdy-pywalfox-native-2.7.4/bin/pywalfox
```

however this isn't version 2.8.0rc1 (I'm on unstable so it should be, right?)
and it also fails to fetch pywal colours with the error `Fetch failed: You are not connected to the Pywalfox daemon`

If I use a nixpkgs overlay like so I explained here to bump it to 2.8.0rc1 (https://github.com/Frewacom/pywalfox/issues/68#issuecomment-2567601729) is still fails to work with the error shown there. Am I doing something wrong?

**Comment by tsandrini** [comment]
@fxzzi Hi!

I'm not super familiar with librewolf, so please correct me if anything is wrong. Anyway, I tried installing librewolf on two of my systems (I haven't ever used it before on these systems) and used the current version of `pywalfox-native` (rev `7ecbbb193e6a7dab424bf3128adfa7e2d0fa6ff9`), installed the pywalfox plugin from the store and everything worked. Can you please try it out with this version instead? Also please try it both with `pywalfox install` and `pywalfox install --browser librewolf`.

Regarding the version, what you refer to `2.8.0rc1` isn't a real tag, it seems to be a hotfix pushed directly into a testing branch of pypi without any versioning in git. nixpkgs is a source based repository, meaning that if there is a source we will always use that instead of other seemingly random tarballs. That is why `2.8.0rc1` isn't here in nixpkgs. The pywalfox-native messenger repo isn't super active unfortunately. Judging from the recursive diff between nixpkgs master and `2.8.0rc1`

```diff
< DAEMON_VERSION = '2.8.0rc1'
---
> DAEMON_VERSION = '2.8.0'
19a20
> BIN_PATH_WIN = os.path.join(APP_PATH, 'bin/win.bat')
diff -r pywalfox-2.8.0rc1/pywalfox/fetcher.py pywalfox-native/pywalfox/fetcher.py
5a6,10
> def create_error(err):
>     error_message = 'Failed to read colors: {0}'.format(err)
>     logging.error(error_message)
>     return (False, None, error_message)
>
32,36c37,42
<
<     except (IOError, JSONDecodeError):
<         error_message = 'Could not read colors from: %s' % PYWAL_COLORS_PATH
<         logging.error(error_message)
<         return (False, None, error_message)
---
>     except JSONDecodeError as err:
>         return create_error(err)
>     except IOError as err:
>         return create_error(err)
>     except:
>         return create_error(sys.exc_info()[1])
diff -r pywalfox-2.8.0rc1/pywalfox/install.py pywalfox-native/pywalfox/install.py
6c6
< from .config import APP_PATH, HOME_PATH
---
> from .config import APP_PATH, HOME_PATH, BIN_PATH_WIN
176c176
< def win_setup(manifest_path_key, bin_path):
---
> def win_setup(manifest_path_key):
195c195
<     copy_manifest(MANIFEST_TARGET_PATH_WIN, bin_path)
---
>     copy_manifest(MANIFEST_TARGET_PATH_WIN, BIN_PATH_WIN)
224c224
<         win_setup(manifest_path_key, bin_path)
---
>         win_setup(manifest_path_key)
diff -r pywalfox-2.8.0rc1/pywalfox/__main__.py pywalfox-native/pywalfox/__main__.py
63d62
<         default=EXECUTABLE_PATH,
132a132,136
> def no_executable():
>     print('Could not find executable path, please re-run the installation with:')
>     print('pywalfox install --executable <path-to-pywalfox-executable>')
>     sys.exit(1)
>
167,170c171
<
<         if not args.custom_path:
<             path_to_use = EXECUTABLE_PATH
<         else:
---
>         if args.custom_path:
171a173,181
>         else:
>             # Try to find a valid path
>             if not os.path.isabs(path_to_use):
>                 path_to_use = os.path.join(
>                     os.path.dirname(os.path.abspath(__file__)),
>                     os.path.basename(__file__)
>                 )
>             if not os.path.isabs(path_to_use) or "__main__.py" in path_to_use:
>                 no_executable()
174,176c184
<             print('Could not find executable path, please re-run the installation:')
<             print('pywalfox install --executable <path-to-pywalfox-executable>')
<             sys.exit(1)
---
>             no_executable()
```

It seems like the nixpkgs master should be newer and cover this case as well + this version has some newer guix related fixes, hence it indeed should be the most recent one, so please report if it works for you, otherwise I can patch it on our side. Thank you.

**Comment by fxzzi** [comment]
Hi there!

Interesting to see it works for you, I reverted all of my overlays and returned back to the latest nixpkgs version 2.7.4. I ran `pywalfox --browser librewolf install` which returned successful:

```
Using executable path: /nix/store/djix9cnscraack2iawpindz9lfj27mdy-pywalfox-native-2.7.4/bin/pywalfox
Copied manifest to: /home/faaris/.librewolf/native-messaging-hosts/pywalfox.json
Set executable path in manifest to: /nix/store/djix9cnscraack2iawpindz9lfj27mdy-pywalfox-native-2.7.4/bin/pywalfox
```

My librewolf is configured in home manager like so:

```nix
    programs.librewolf = {
      enable = true;
      package = pkgs.librewolf.overrideAttrs (_old: {
        extraNativeMessagingHosts = with pkgs; [pywalfox-native];
      });
      languagePacks = [
        "en-GB"
        "en-US"
      ];
    };
```

However trying to use the pywalfox extension gives me this error: 

![Image](https://github.com/user-attachments/assets/bfe9c3e1-4656-4ff1-ac1e-fe7b63223650)

Is there an issue in my configurations?

**Comment by tsandrini** [comment]
@fxzzi No problem! Apologies for the late reply I was a bit more busy with work lately unfortunately. Anyway, glad it worked and I also played around with a completely blank VM with just a basic librewolf setup + with your snippet included as well. A few points regarding that

1. This line is not needed since we are running the `pywalfox install` manually and it could even potentially inhibit how it currently works. I could automate this such that we wouldn't need to run the `pywalfox install` at all, but that would require a separate NixOS/HM module for pywalfox, which I am not sure how useful it would be aside from just this single purpose.
```nix
      package = pkgs.librewolf.overrideAttrs (_old: {
        extraNativeMessagingHosts = with pkgs; [pywalfox-native];
      });
```
2. No matter what configuration I tried, `pywalfox install --browser librewolf` didn't work. This explains why it worked previously on my machine, it was reading the `.mozilla/native-messaging-hosts/pywalfox.json` file from within my firefox config. 
3. If a fallback `.mozilla/native-messaging-hosts/pywalfox.json` is present, pywalfox works as expected in librewolf.

This is definitely a strange behaviour, since the `pywalfox.json` points to the same executable. Even the rec diff between our main version and the upstream custom pypi patch that I included doesn't really indicate why this might be happening.

 I recommend using the `.mozilla/native-messaging-hosts/pywalfox.json` file for the time being and I guess if there are going to be more people running pywalfox + librewolf on nix I might look deeper into debugging this issue.
